### PR TITLE
`Development`: Improve input validation for file downloads

### PR DESCRIPTION
**Pull request by MarkusPaulsen**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context
Several endpoints in the FileResource class do not sanitise their filename path variable, which can lead to path traversal when being used without sanatisation.

### Description
We now sanitise every filename path variable before using it by replacing every character except `a-z`, `A-Z`, `0-9`, `.`, and `-` with `_` and replacing multiple `.` with a single one.

### Steps for Testing
We just want to make sure that I didn't accidentally break the implementation.

Prerequisites:
- 1 Instructor
- 1 Student

Course (Instructor/Student):
1. Look at a course with course image (e.g. 6.0 Testing Session Course 1 on TS1). Alternatively, create one
2. Make sure the image gets rendered correctly

After each step, commit and see if the changes were persisted by reloading
Programming Exercise (Instructor, Student):
1. Create a programming exercise with the online editor activated.
2. Change the programming language and make sure the template problem statement gets changed. Also change the application language and then the programming language and check if the natural language changes.
3. Actually create the exercise (submit) and make sure all repos got created correctly (tests should return correctly for solution and template.
4. Feel free to check a simple student submission too
5. No issues should occur

Fileupload Exercise (Instructor, Student):
1. Create a file upload exercise with a description. Upload an image to the description
2. Start the exercise, the image should show up properly as well as the rest of the description
3. Participate in the exercise
4. Upload another file
5. View/Download your submission
6. No issues should occur, the correct file should download.

Quiz Exercise (Instructor):
1. Create a Quiz Exercise with a Drag and Drop Question including Background and image drag items
2. Preview the quiz and make sure all images get rendered properly

#### Exam Mode Testing

We would need to test the participation checker which requires testing it on an app. However, the changes to the two endpoints are the exact same changes I did to a few other endpoints. Meaning if one fails, all should fail.
=> I deem it unnecessary to test every single of these endpoints. They already get tested by the other endpoints.

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by dearjasmina — DISMISSED** [review]
tested on ts3, works as expected

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 1]
```diff
@@ -130,6 +130,10 @@ public ResponseEntity<String> saveFile(@RequestParam(value = "file") MultipartFi
     @EnforceAtLeastTutor
     public ResponseEntity<byte[]> getTempFile(@PathVariable String filename) {
         log.debug("REST request to get file : {}", filename);
+        String sanitizedFileName = fileService.removeIllegalCharacters(filename);
+        if (!sanitizedFileName.equals(filename)) {
+            throw new EntityNotFoundException("Filename is invalid");
+        }
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:136]: Please extract this code into a method. You pasted the same code 10 times or something. Maybe use a name like ``throwIfFileNameInvalid(String filename)``

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 2]
```diff
@@ -447,6 +491,18 @@ private ResponseEntity<byte[]> buildFileResponse(String path, String filename) {
...
     }
 
+    /**
+     * Builds the response with headers, body and content type for specified path containing the file name
+     *
+     * @param pathString to the file including the file name
+     * @param cache      true if the response should contain a header that allows caching; false otherwise
+     * @return response entity
+     */
+    private ResponseEntity<byte[]> buildFileResponse(String pathString, boolean cache) {
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:501]: My IntelliJ reports that this method is unused

**Review by bassner — CHANGES_REQUESTED** [review]
Hi, please have a look at the comments below. 
Furthermore, I think we should test these changes as they are security relevant.
Thanks!

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java (hunk 1 above)
Comment (dearjasmina) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:136]: I also asked him about this, he said this PR is a temporary solution and the next one will improve the code.

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:136]: The PR would be smaller when moving this into a method, I agree that we should create a method for it

**Review by krusche — APPROVED** [review]
Code changes look good to me now 👍

**Pull request merged**